homeassistant/components/generic/config_flow.py

```diff
@@ -0,0 +1,149 @@
+"""Config flow for generic (IP Camera)."""
+
+import imghdr
+import logging
+from typing import Any, Dict
+
+import av
+import voluptuous as vol
+
+from homeassistant import config_entries, data_entry_flow
+from homeassistant.components.camera import DEFAULT_CONTENT_TYPE
+from homeassistant.const import (
+    CONF_AUTHENTICATION,
+    CONF_NAME,
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+    HTTP_BASIC_AUTHENTICATION,
+    HTTP_DIGEST_AUTHENTICATION,
+)
+
+from .camera import GenericCamera
+
+# pylint: disable=unused-import
+from .const import (
+    CONF_CONTENT_TYPE,
+    CONF_FRAMERATE,
+    CONF_LIMIT_REFETCH_TO_URL_CHANGE,
+    CONF_STILL_IMAGE_URL,
+    CONF_STREAM_SOURCE,
+    DEFAULT_NAME,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class GenericIPCamConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for generic IP camera."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialise the config flow."""
+        super().__init__()
+        self._errors = {}
+
+    async def _test_connection(self, info):
+        """Verify that the camera data is valid before we add it."""
+        await self.async_set_unique_id(info[CONF_STILL_IMAGE_URL])
+        try:
+            self._abort_if_unique_id_configured()
+        except data_entry_flow.AbortFlow as err:
+            if "already_configured" in str(err):
+                self._errors["base"] = "still_image_already_configured"
+                return False
+        cam = GenericCamera(self.hass, info)
+        # Minimum functionality is getting a still image
+        image = await cam.async_camera_image()
+        if image is None:
+            self._errors["base"] = "unable_still_load"
+            return False
+        fmt = imghdr.what(None, h=image)
+        _LOGGER.debug(
+            "Still image at '%s' detected format: %s", info[CONF_STILL_IMAGE_URL], fmt
+        )
+        if fmt is None:
+            self._errors["base"] = "invalid_still_image"
+            return False
+
+        # Second level functionality is to get a stream.
+        if info.get(CONF_STREAM_SOURCE) not in [None, ""]:
+            try:
+                container = av.open(info.get(CONF_STREAM_SOURCE), options=None)
```

Comment:
`av.open` is synchronous I/O, which is why it's wrapped in a thread in the `stream compoenent`. Here, since you are just ensuring that there is a video stream, you can pull this out into a separate method and [run it in the executor](https://developers.home-assistant.io/docs/asyncio_working_with_async/#calling-sync-functions-from-async).